File: 739-daily-temperatures/daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> answer(n, 0);
        stack<int> stk;  // stores indices

        for (int i = 0; i < n; i++) {
            while (!stk.empty() && temperatures[stk.top()] < temperatures[i]) {
                int prevIdx = stk.top();
                stk.pop();
                answer[prevIdx] = i - prevIdx;
            }

            stk.push(i);
        }

        return answer;
    }
};

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        n = len(temperatures)
        answer = [0] * n
        stack = []  # stores indices, monotonic decreasing temperatures

        for i in range(n):
            # While current temp is warmer than the temp at stack's top index
            while stack and temperatures[stack[-1]] < temperatures[i]:
                prev_idx = stack.pop()
                answer[prev_idx] = i - prev_idx  # days waited

            stack.append(i)

        return answer